### ISSUE TITLE
Allow players to win stale games by forfeit

### ISSUE THREAD

**Issue by irilyth** [buttonmen-dev/buttonmen#1682]
If one player hasn't moved in a game for a while (two weeks?), the other player should get the choice to win by forfeit.

**Comment by irilyth** [comment]
We had previously thought that this wouldn't be all that useful until we had tournaments, but it'd be somewhat useful in Mad Robin and UBFC too.

**Comment by irilyth** [comment]
It'd also be useful for people who accept games from new players, who tend to wander off.

**Comment by dwvanstone** [comment]
I would encourage that we store something internally that marks the win as Win By Forfeit so the button stats can ignore these games.

**Comment by irilyth** [comment]
Definitely. I also suspect that there'll be different use cases for this: For example, a forfeit is the same as a loss for UBFC purposes, so I'd like it *not* to be implemented as something like "the game is retroactively deleted as if it never existed" or something (which is sort of how I recall it working on the old site).

**Comment by AdmiralJota** [comment]
One option would be to keep the current behavior (i.e., let the game remain stale forever), but once a game goes stale, display that fact on the game page, and allow things like tournaments to *treat* the stale game like a completed one, even though it's technically not.

**Comment by irilyth** [comment]
I think I'd rather differentiate between stale and forfeits; among other things, I like that forcing a forfeit requires action from the other player, so if they know of some extenuating circumstances, or just don't want to end the game, they can make that decision. I like that better than having it be purely clock based.

Also, what would happen if the tournament ignores it, and moves on as if it had ended, but then the other player comes back and resumes it? Would something indicate that the game was out of the tournament or whatever?

That all sounds needlessly complicated -- I just think adding an extra bit to say "this game was forfeited" lets you distinguish between them and any other completed game. What's a use case where that *wouldn't* be sufficient, or where something else would have some advantages?

**Comment by AdmiralJota** [comment]
Is it fair to the other tournament players to have the game sit indefinitely because one player doesn't want to forcibly end the game? Or if both players go idle?

(And I think adding a comment to the game log indicating that it was forfeited out of the tournament seems straightforward enough.)

I'm not saying that I'm opposed to players force forfeits from idle opponents. I'm just positing some responses to the points you've made.

**Comment by irilyth** [comment]
That used to happen sometimes on the old site, and I agree it's something we should try to avoid. Here's a proposal: After N days without action, a game becomes eligible for forfeit, at which point any of these people can trigger a forfeit loss to the idle player:

* The idle player. Perhaps they've given up on the site, have come back one last time to clean up, and would like to forfeit the whole game, not just surrender rounds and otherwise slog through the rest of the game. Perhaps uncommon, but why not allow it.
* The other player. This is probably the most common case.
* The tournament organizer, if the game is part of a tournament.
* The person who created the game, if that's different from the players. This wouldn't have any effect until we allow third-party game creation (#1784), but when we do, it'd be nice if I (as e.g. UBFC or Mad Robin wrangler) could forfeit games that have gone idle. (This is basically equivalent to "the tournament organizer", but for games that aren't part of formal tournaments.)

Whatcha think?

**Comment by cgolubi1** [comment]
Irilyth's proposal seems reasonable to me.  I think we can always expand/change the set of players who can forfeit a game later, so the initial value of that access control list doesn't strike me as fundamental to the proposal (and, again, the proposed set seems fine as a starting point).

For me, the two main design features i want to see are:
* There should be some way to move forward from a game which is stuck (whether that's by letting a player win by forfeit, or by having the context which cares about the game (e.g. a tournament) be able to declare a forfeit outcome separate from whether the actual underlying game is actually over)
* It should not be *mandatory* for stale games to be forfeited outside of tournaments --- my stale games are nicely hidden on my Overview, and i'd just as soon keep them all around in case any of those players come back.

One question i have is: why are we talking about this before we have tournaments?  That is, what's the use case for forfeiting stale games right now (when we don't have tournaments or third party game creation or any of those things)?  Or are we just planning for tournaments?

**Comment by irilyth** [comment]
I mentioned two recently: Mad Robin and UBFC, and people who accept games from new players, who tend to wander off and never come back at a somewhat above-average rate.

Hiding the stale games may be fine for the second one -- I thought someone said something on the forums that made me think of that, but I can't find it now. The Mad Robin and UBFC part would make things smoother for me even today.

**Comment by blackshadowshade** [comment]
Also adding http://www.buttonweavers.com/ui/forum.html#!threadId=151, where people also expressed the wish to delete stale games.

**Comment by blackshadowshade** [comment]
My thought here is to allow a player to change the game state to CANCELLED after it has been idle for 60 days. That would give neither player a win in the stats and remove it from the Overview page, but leave all other data unchanged, including round wins that have already occurred.

**Comment by blackshadowshade** [comment]
However, this might have other ramifications, so it might be safer to put in a new game state ABANDONED, which will allow us to customise the behaviour of various game screens as necessary.

**Comment by blackshadowshade** [comment]
I've had another think about this, due to a bump on the forum thread by tasha.

One overriding reason for wanting to be able to "delete a stale game" appears to be a wish to clean up the stale game section of the overview page.

We could accommodate this by adding the ability to unfollow / hide a specific stale game. In practice, this would require an extra 'hide' boolean column in game_player_map. Then the Overview page would show all games except those that had a true value in the 'hide' column *and* were stale by 14 days or more due to the opponent's inactivity.

What do you think of this idea?

Progressing tournaments will happen separately, I intend to add the ability for tournament owners or game participants to 'bump' a tournament or game to update its tournament status.

**Comment by cgolubi1** [comment]
In principle, i like that idea; it seems like it meets the player ask without any changes which could cause surprising problems.

Implementation-wise: you'd need to walk through the logic of what happens when the other player takes a move on a hidden game.  My recommendation is that we build this logic:
* The "hide" API call fails if the game is not stale by 14 days or more --- UI or not, you can't set the hide flag on a game which isn't stale
* Activity on a game *clears* any true values from the 'hide' column --- again, once the game becomes not-stale, it no longer has a hide flag

If we build logic that ensures that only actually-stale-right-now games can have a hide flag set, we get the following nice things:
* There's no risk of a bug which hides games in the top part of the overview, because active games can't have the hide flag set.
* It simplifies the question of what happens when a hidden game became briefly active and then stale again - the game is no longer hidden, and if the player wants to hide it again, they have to make that decision again.  (This seems correct because the game *was* active for awhile, so the player may decide differently the second time.)

**Comment by blackshadowshade** [comment]
That matches my understanding of the solution, yes.

**Comment by cgolubi1** [comment]
Oh, i guess the other implementation thing i want to note is that this is another case where we could wind up building the same logic in both the UI and the API, but we shouldn't; what we should *actually* do is have the game listing APIs return a new `isHideable` flag with each game, and have the UI use that in deciding whether to provide the "Hide Game" button/link/box.

[I think we've all learned this by this point, but it's also been awhile, so i figured i'd remind us.]

**Comment by blackshadowshade** [comment]
Thanks for the reminder, that's definitely a good idea.
